src/segment-loader.js

```diff
@@ -2044,34 +2063,138 @@ export default class SegmentLoader extends videojs.EventTarget {
     return null;
   }
 
-  appendToSourceBuffer_({ segmentInfo, type, initSegment, data }) {
-    const segments = [data];
-    let byteLength = data.byteLength;
+  handleQuotaExceededError_({segmentInfo, type, bytes}, error) {
+    const audioBuffered = this.sourceUpdater_.audioBuffered();
+    const videoBuffered = this.sourceUpdater_.videoBuffered();
 
-    if (initSegment) {
-      // if the media initialization segment is changing, append it before the content
-      // segment
-      segments.unshift(initSegment);
-      byteLength += initSegment.byteLength;
+    // For now we're ignoring any notion of gaps in the buffer, but they, in theory,
+    // should be cleared out during the buffer removals. However, log in case it helps
+    // debug.
+    if (audioBuffered.length > 1) {
+      this.logger_('On QUOTA_EXCEEDED_ERR, found gaps in the audio buffer: ' +
+        timeRangesToArray(audioBuffered).join(', '));
+    }
+    if (videoBuffered.length > 1) {
+      this.logger_('On QUOTA_EXCEEDED_ERR, found gaps in the video buffer: ' +
+        timeRangesToArray(videoBuffered).join(', '));
     }
 
-    // Technically we should be OK appending the init segment separately, however, we
-    // haven't yet tested that, and prepending is how we have always done things.
-    const bytes = concatSegments({
-      bytes: byteLength,
-      segments
+    const audioBufferStart = audioBuffered.length ? audioBuffered.start(0) : 0;
+    const audioBufferEnd = audioBuffered.length ?
+      audioBuffered.end(audioBuffered.length - 1) : 0;
+    const videoBufferStart = videoBuffered.length ? videoBuffered.start(0) : 0;
+    const videoBufferEnd = videoBuffered.length ?
+      videoBuffered.end(videoBuffered.length - 1) : 0;
+
+    if (
+      (audioBufferEnd - audioBufferStart) <= MIN_BACK_BUFFER &&
+      (videoBufferEnd - videoBufferStart) <= MIN_BACK_BUFFER
+    ) {
+      // Can't remove enough buffer to make room for new segment (or the browser doesn't
+      // allow for appends of segments this size). In the future, it may be possible to
+      // split up the segment and append in pieces, but for now, error out this playlist
+      // in an attempt to switch to a more manageable rendition.
+      this.logger_('On QUOTA_EXCEEDED_ERR, single segment too large to append to ' +
+        'buffer, triggering an error. ' +
+        `Appended byte length: ${bytes.byteLength}, ` +
+        `audio buffer: ${timeRangesToArray(audioBuffered).join(', ')}, ` +
+        `video buffer: ${timeRangesToArray(videoBuffered).join(', ')}, `);
+      this.error({
```

Comment:
I wonder if we should try splitting the single append into two appends and halving the segment.

Reply:
Yeah, I added that as a comment above. I think it would be good to try as a TODO, but I'm hesitant to do it just yet as I'm not sure how different browsers will behave if we append bytes that aren't split up on proper box boundaries.